mantidproject/mantid
# Muon analysis add period column to group table

Merge after #29135

**Description of work.**

There is a scientific need to be able to specify for each group which periods you want to use to compose that group. This PR adds a grouping column to the groups table to allow this. It has also been updated to automatically default to creating a group/pair for each period for multiperiod data.

There are several edge cases concerning the loading of runs with different numbers of periods at the same time. In general this is allowed and we do the best we can going forward creating groups and pairs for each run based on the periods it does have. The one place this does not currently work is for the sequential simultaneous fitting of runs with different periods. This will currently give a warning and not work. As this is a rare edge case which resolves gracefully and would take substantial effort to fix I have left it in for now.

The full list of requirements in the issue which should be check ed were:

It should handle:

Period not existing (no crash or warning)
Generate the appropriate data
Can allow for periods to be summed (e.g.typing 1,2 or 1-3)
Add the period names to the run info on the home tab
Allow the user to not have an instance of a period (e.g. fwd for periods 1 and 3)
Clicking default will only produce the groups for the periods that are present in any of the runs.

Of these all are done except for, Add the period names to the run info on the home tab, which I left for a separate PR as it is easily separable and this changeset is already quite large.





**To test:**

You will need access to the ISIS data archive to test this PR. 

 * Open the Muon Analysis GUI load HIFI84447 (this is a run with 4 periods)
 * Check that a group and pair is created for each period.
 * Check that multiple periods can be specified for each group.
 * Load HIFI44447 in addition, this is a single period run. Check that the correct groups and pairs are produced for it.
 * Click default and check that only groups/pairs for period 1 are produced.
 * Try removing some groups and checking everything is handled gracefully.
 * The above is mostly sunny day cases so have a play around and see if there is any strange behaviiur for odd action sequences.
 



Fixes #29114  




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
